### ISSUE TITLE
Upgrade issue - 'yo jhipster:upgrade' failed from 4.5.1 to 4.5.2

### ISSUE THREAD

**Issue by vadym-golub-axon** [jhipster/generator-jhipster#5869]
##### **Overview of the issue**
Executing 'yo jhipster:upgrade' on the first upgrade of the project generates `jhipster_upgrade` branch using the latest version.

Here is the output:
```
Welcome to the JHipster Upgrade Sub-Generator
This will upgrade your current application codebase to the latest JHipster version
Looking for latest generator-jhipster version...
New generator-jhipster version found: 4.5.2
Git repository detected
Created branch jhipster_upgrade
Cleaned up directory
Installing JHipster 4.5.1 locally
Installed generator-jhipster@4.5.1
Regenerating app with jhipster 4.5.1...
Successfully regenerated app with jhipster 4.5.1
Committed with message "Generated with JHipster 4.5.1"
Checked out branch "HN-46"
Current code recorded as generated with version 4.5.1
Updating generator-jhipster to 4.5.2 . This might take some time...
Updated generator-jhipster to version 4.5.2
Error jhipster:upgrade 

ERROR! Unable to checkout branch jhipster_upgrade:
error: Your local changes to the following files would be overwritten by checkout:
	prk/package.json
Please commit your changes or stash them before you switch branches.
Aborting
```
after these actions, the newly created commit has the name 'Generated with JHipster 4.5.1' and contains README.md files that were generated with 4.5.2 version and also package.json and .yo-rc.json contain wrong version of jhipster('4.5.2')

package.json (jhipster_upgrade branch)
```
{
  "devDependencies": {
    "generator-jhipster": "4.5.2"
  }
}
```
.yo-rc.json (jhipster_upgrade branch)
```
{
  "generator-jhipster": {
   ...
    "jhipsterVersion": "4.5.2"
   ...
  }

}
```

package.json causing conflict of the upgrading branch
```
{
  "devDependencies": {
    "generator-jhipster": "4.5.2"
  }
}
```

The problem with uncommitted changes is described in issue #5637 can be resolved by committing package.json, but as the project was generated in jhipster_upgrade branch with 4.5.2 version therefore after comletion of upgrade jhipster adds new commit with the name 'Generated with JHipster 4.5.2' which doesn't have a diff with  'Generated with JHipster 4.5.1' and no changes are added to upgrading branch.

##### **Related issues**
#5637

##### **JHipster Version(s)**

```
/home/vadim/HN/server2/prk
└── generator-jhipster@4.5.2 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

```yaml
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "hapi.prk",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "4.5.1",
    "baseName": "prk",
    "packageName": "hapi.prk",
    "packageFolder": "hapi/prk",
    "serverPort": "8081",
    "authenticationType": "jwt",
    "hibernateCache": "hazelcast",
    "clusteredHttpSession": false,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "postgresql",
    "prodDatabaseType": "postgresql",
    "searchEngine": "elasticsearch",
    "messageBroker": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "maven",
    "enableSocialSignIn": false,
    "jwtSecretKey": "e7484ae1f85b037250854642cee6d4481c0a2520",
    "enableTranslation": true,
    "applicationType": "microservice",
    "testFrameworks": [
      "gatling",
      "cucumber"
    ],
    "jhiPrefix": "jhi",
    "skipClient": true,
    "skipUserManagement": true,
    "nativeLanguage": "en",
    "languages": [
      "en",
      "es",
      "tr"
    ],
    "clientPackageManager": "yarn"
  }
}
```

##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

ls: no such file or directory: .jhipster/*.json

##### **Browsers and Operating System**

java version "1.8.0_121"
Java(TM) SE Runtime Environment (build 1.8.0_121-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.121-b13, mixed mode)

git version 2.11.0

node: v6.9.5

npm: 3.10.10

bower: 1.3.12

gulp:
[15:26:37] CLI version 1.3.0

yeoman: 1.8.5

yarn: 0.24.5

Docker version 1.13.1, build 092cba3

docker-compose version 1.9.0, build 2585387

Execution complete

- [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by jdubois** [comment]
Yes we already have 2 issues with this currently opened. Unfortunately we lack time to solve this.

**Comment by jdubois** [comment]
Ping @lordlothar99 as he coded most of this. I'll have a look this afternoon.

**Comment by hugotown** [comment]
@jdubois 
That is an important feature of JHipster!
some we need upgrade our applications, I hope to do it soon...

Best Regards!

**Comment by jdubois** [comment]
I know it's important, but it's really complex. My current guess is that the first project regeneration fails (the one with the older version), and that depends how you installed JHipster (I have one system where it fails, another one where it works...). It would be when the upgrade script cannot upgrade the generator, and in that case you're stuck with the same version all the time.

**Comment by tezcane** [comment]
Here is how I usually upgrade my project (I don't use jhipster:upgrade):
1. yo jhipster
2. accept all changes
3. diff new changes to see new changes plus merge with my customization

This should get around the problem...

**Comment by jdubois** [comment]
Closed by #5883
